src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs

```diff
@@ -1361,53 +1369,61 @@ internal bool Compile(bool optimized)
 
         internal static void LogScriptBlockCreation(ScriptBlock scriptBlock, bool force)
         {
+            if (scriptBlock.HasLogged && !InternalTestHooks.ForceScriptBlockLogging)
+            {
+                // Fast exit if the script block is already logged and we are not force re-logging in tests.
+                return;
+            }
+
             ScriptBlockLogging logSetting = GetScriptBlockLoggingSetting();
             if (force || logSetting?.EnableScriptBlockLogging == true)
             {
-                if (!scriptBlock.HasLogged || InternalTestHooks.ForceScriptBlockLogging)
+                // If script block logging is explicitly disabled, or it's from a trusted
+                // file or internal, skip logging.
+                if (logSetting?.EnableScriptBlockLogging == false ||
+                    scriptBlock.ScriptBlockData.IsProductCode)
                 {
-                    // If script block logging is explicitly disabled, or it's from a trusted
-                    // file or internal, skip logging.
-                    if (logSetting?.EnableScriptBlockLogging == false ||
-                        scriptBlock.ScriptBlockData.IsProductCode)
-                    {
-                        return;
-                    }
+                    scriptBlock.SkipLogging = true;
```

Comment:
It seems we skip logging (with line 1425) only if
```powershell
if (scriptBlock.ScriptBlockData.IsProductCode)
{
    scriptBlock.SkipLogging = true;
}
```
We could simplify the code and move this to line 1379.